File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.java
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        Deque<TreeNode> stack = new ArrayDeque<>();
        TreeNode curr = root;

        while (!stack.isEmpty() || curr != null) {
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }
            curr = stack.pop();
            k--;
            if (k == 0) return curr.val;
            curr = curr.right;
        }

        return -1;
    }
}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-2.cpp
class Solution {
public:
    int preIdx = 0;
    vector<int> preorder;
    unordered_map<int, int> inMap;

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        this->preorder = preorder;
        for (int i = 0; i < inorder.size(); i++) {
            inMap[inorder[i]] = i;
        }
        return build(0, inorder.size() - 1);
    }

    TreeNode* build(int l, int r) {
        if (l > r) return nullptr;

        int rootVal = preorder[preIdx++];
        TreeNode* root = new TreeNode(rootVal);

        int mid = inMap[rootVal];
        root->left = build(l, mid - 1);
        root->right = build(mid + 1, r);

        return root;
    }
};